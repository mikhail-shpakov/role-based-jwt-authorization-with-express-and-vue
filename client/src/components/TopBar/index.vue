<template lang="pug">
.section.is-vertical-center
  b-navbar.navbar-custom
    template(slot="brand")
      b-navbar-item.item-logo(
        href="https://shpakov.dev/"
        target="_blank"
        rel="noopener"
      )
        img(
          src="@/assets/logo.svg"
          alt="img-logo"
        )

    template(slot='start')
      b-button.is-text(
        tag="router-link"
        to="Servers"
        v-if="accessToken"
      ) Сервера

    template(slot='end')
      b-button.is-text(
        tag="router-link"
        to="About"
      ) О приложении
      b-button(
        tag="a"
        href="https://github.com/mikhail-shpakov/role-based-jwt-authorization-with-express-and-vue"
        target="_blank"
        rel="noopener"
        icon-left="github"
      ) Репозиторий
      b-button.is-danger.is-outlined(
        tag="a"
        @click="logout"
        :loading="isLoading"
        v-if="accessToken"
      ) Выйти
</template>

<script>
import { mapState, mapActions } from 'vuex'

export default {
  name: 'TopBar',
  data () {
    return {
      isLoading: false
    }
  },
  computed: {
    ...mapState('user', {
      accessToken: state => state.tokens.accessToken
    })
  },
  methods: {
    ...mapActions('user', [
      'LOGOUT'
    ]),
    async logout () {
      this.isLoading = true
      await this.LOGOUT()
      this.isLoading = false
    }
  }
}
</script>

<style lang="sass" scoped>
div.section
  border-bottom: 1px solid $color-alt
  padding: 0
  vertical-align: center
  @media (min-width: $display-bp-desktop)
    padding: 0.75rem 0

    .navbar-custom
      width: $display-width
      margin: auto

  .navbar-item img
    max-height: 36px
    @media (min-width: $display-bp-desktop)
      max-height: 50px

  a.button
    align-self: center
    margin: 0.5rem 0.75rem
    display: table
    width: auto
    @media (min-width: $display-bp-desktop)
      margin: 0 0 0 0.75rem
      display: inline-flex

    &.is-text
      text-decoration: none
</style>
