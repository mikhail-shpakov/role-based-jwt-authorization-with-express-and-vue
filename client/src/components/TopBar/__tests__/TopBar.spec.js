import { createLocalVue, shallowMount } from '@vue/test-utils'
import Vuex from 'vuex'
import Buefy from 'buefy'
import TopBar from '../index'

const localVue = createLocalVue()
localVue.use(Buefy)
localVue.use(Vuex)

describe('TopBar', () => {
  let store
  let actions

  beforeEach(() => {
    const state = {
      tokens: {
        accessToken: 'some token'
      }
    }

    actions = {
      LOGOUT: jest.fn()
    }

    store = new Vuex.Store({
      modules: {
        user: {
          state,
          actions,
          namespaced: true
        }
      }
    })
  })

  it('when load page and user non auth, TopBar renders correctly', () => {
    store.state.user.tokens.accessToken = ''

    const wrapper = shallowMount(TopBar, {
      localVue,
      store
    })
    const buttonLogout = wrapper.find({ ref: 'button-logout' })
    const buttonLink = wrapper.find({ ref: 'button-link' })

    expect(buttonLogout.exists()).toEqual(false)
    expect(buttonLink.text()).toEqual('Войти')
    expect(wrapper).toMatchSnapshot()
  })

  it('when load page and user auth, TopBar renders correctly', () => {
    const wrapper = shallowMount(TopBar, {
      localVue,
      store
    })
    const buttonLogout = wrapper.find({ ref: 'button-logout' })
    const buttonLink = wrapper.find({ ref: 'button-link' })

    expect(buttonLogout.exists()).toEqual(true)
    expect(buttonLink.text()).toEqual('Сервера')
    expect(wrapper).toMatchSnapshot()
  })

  it('when user auth and click logout Button, will be called Vuex LOGOUT()', () => {
    const wrapper = shallowMount(TopBar, {
      localVue,
      store
    })
    const button = wrapper.find({ ref: 'button-logout' })

    button.trigger('click')

    expect(actions.LOGOUT).toHaveBeenCalled()
  })
})
