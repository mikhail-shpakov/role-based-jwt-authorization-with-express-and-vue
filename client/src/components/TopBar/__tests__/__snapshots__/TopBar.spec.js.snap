// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopBar when load page and user auth, TopBar renders correctly 1`] = `
<div class="section is-vertical-center">
  <b-navbar-stub fixedtop="true" closeonclick="true" mobileburger="true" class="navbar-custom"><template></template><template></template><template></template></b-navbar-stub>
</div>
`;

exports[`TopBar when load page and user non auth, TopBar renders correctly 1`] = `
<div class="section is-vertical-center">
  <b-navbar-stub fixedtop="true" closeonclick="true" mobileburger="true" class="navbar-custom"><template></template><template></template><template></template></b-navbar-stub>
</div>
`;
